File: ga/glm4_result_2024_6/result/1968G2_cpp_glm/30_1968G2.cpp
use std::io::{self, BufRead};
use std::cmp::Ordering;

const N: usize = 2e5 as usize + 3;
const MAX_N: usize = N + 1;

fn find(p: &mut [usize], k: usize) -> usize {
    if k > N { return k; }
    if p[k] == k { return k; }
    p[k] = find(p, p[k]);
    p[k]
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();
    
    let mut t = lines.next().unwrap().unwrap().parse::<usize>().unwrap();
    while t > 0 {
        t -= 1;
        
        let mut input = lines.next().unwrap().unwrap().split_whitespace();
        let n = input.next().unwrap().parse::<usize>().unwrap();
        let x = input.next().unwrap().parse::<usize>().unwrap();
        let y = input.next().unwrap().parse::<usize>().unwrap();
        let a = input.next().unwrap();
        
        let mut p = vec![0; MAX_N];
        let mut z = vec![0; MAX_N];
        let mut ans = vec![0; MAX_N];
        let mut t = (1..=n).collect::<Vec<usize>>();
        
        z[1] = n;
        for i in 2..=n {
            let mut j = 0;
            while j + z[j] > i { j = i - j - z[j] + 1; }
            while a.chars().nth(i + z[i] - 1) == a.chars().nth(1 + z[i] - 1) { z[i] += 1; }
            if i + z[i] > j + z[j] { j = i; }
        }
        
        t.sort_by(|&i, &j| z[i].cmp(&z[j]));
        
        for i in 1..=n {
            let mut k = 1;
            while k <= n && z[t[k]] < i { p[t[k]] = t[k] + 1; k += 1; }
            let mut c = 0;
            for j in 1..=n { c += 1; while find(&mut p, j + i) != j { j = find(&mut p, j + i); } }
            ans[c] = i;
        }
        
        for i in (1..=n).rev() { ans[i] = ans[i].max(ans[i + 1]); }
        
        for i in x..=y { print!("{} ", ans[i]); }
        println!("");
    }
}

File: ga/gpt3_5_result_2024_6/result/1968G2_gpt/30_1968G2.rs
use std::cmp;

const N: usize = 200000;
const N_: usize = 1 << 18; /* N_ = pow2(ceil(log2(N))) */

fn max(a: usize, b: usize) -> usize {
    if a > b {
        a
    } else {
        b
    }
}

fn build(aa: &[usize], n: usize) -> Vec<usize> {
    let mut st = vec![0; N_ * 2];
    let mut n_ = 1;

    while n_ < n {
        n_ <<= 1;
    }

    for i in 0..n {
        st[n_ + i] = aa[i];
    }

    for i in (1..n_).rev() {
        st[i] = max(st[i << 1], st[i << 1 | 1]);
    }

    st
}

fn query(l: usize, a: usize, n_: usize, st: &Vec<usize>) -> usize {
    let mut l = l;
    let mut r = n_ - 1;

    while l <= r {
        if l & 1 == 1 {
            if st[l] >= a {
                let mut temp_l = l;
                while temp_l < n_ {
                    temp_l = if st[temp_l << 1] >= a {
                        temp_l << 1
                    } else {
                        temp_l << 1 | 1
                    };
                }

                return temp_l - n_;
            }

            l += 1;
        }

        l >>= 1;
        r >>= 1;
    }

    return n_;
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).expect("Failed to read input");
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).expect("Failed to read input");
        let mut iter = input.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let kl: usize = iter.next().unwrap().parse().unwrap();
        let kr: usize = iter.next().unwrap().parse().unwrap();
        let cc: String = iter.next().unwrap().parse().unwrap();

        let mut zz = vec![0; n];
        let mut l = 0;
        let mut r = 0;

        for i in 1..n {
            if i + zz[i - l] < r {
                zz[i] = zz[i - l];
            } else {
                l = i;
                r = cmp::max(r, i);

                while r < n && cc.as_bytes()[r] == cc.as_bytes()[r - l] {
                    r += 1;
                }

                zz[i] = r - l;
            }
        }

        let st = build(&zz, n);
        let mut ans = vec![0; n + 1];

        for p in 1..=n {
            let mut k = 0;
            let mut i = 0;

            while i < n {
                i = query(i + p, p, n, &st);
                if i < n {
                    k += 1;
                }
            }

            ans[k] = cmp::max(ans[k], p);
        }

        for k in (1..n).rev() {
            ans[k] = cmp::max(ans[k], ans[k + 1]);
        }

        for k in kl..=kr {
            print!("{} ", ans[k]);
        }

        println!();
    }
}